File: java/core/src/main/java/dev/varar/core/Scanner.java
package dev.varar.core;

import dev.varar.core.Ast.Block;
import dev.varar.core.Ast.Row;
import dev.varar.core.Ast.SegmentOffset;
import dev.varar.core.TableCells.RowCells;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Turns raw Markdown source into a flat list of {@link Block} nodes: headings, paragraphs, list
 * items, blockquotes, fenced code, thematic breaks, tables.
 *
 * <p>Port of {@code var-core/src/scanner.ts}. Java's {@code String}/{@code char} are already
 * UTF-16 code-unit indexed (see {@link Span}'s javadoc), so — as with the other core modules —
 * this iterates and slices directly on {@code String}/{@code Matcher} offsets, exactly as {@code
 * scanner.ts} does, with no code-point conversion layer. {@code java.util.regex.Pattern}'s {@code
 * ^}/{@code $} anchors and {@code Matcher.start()}/{@code .end()} land on the same UTF-16 units as
 * JS's regex {@code .index}, confirmed here (see {@code astralParagraphSpanEndOffsetIsUtf16CodeUnits}
 * in {@code ScannerTest}) exactly as every prior module found.
 *
 * <p>The {@code plugins} parameter carried by {@code scanner.ts}'s (and the Python port's) {@code
 * scan} signature is intentionally out of scope here — no scanner plugin is needed by this port
 * yet (mirrors the Python port's Task 5 note); {@link #scan} takes no plugins parameter at all.
 */
public final class Scanner {

    private Scanner() {}

    /**
     * A scanner-line: one line of source between (or before/after) {@code \n} boundaries, plus its
     * start/end offsets in the full source. Package-private — internal to the scanner.
     */
    record RawLine(String text, int startOffset, int endOffset) {}

    private static final Pattern THEMATIC_RE = Pattern.compile("^\\s*([-*_])(\\s*\\1){2,}\\s*$");
    private static final Pattern UL_RE = Pattern.compile("^(\\s*)([-*+])\\s+(.*)$");
    private static final Pattern OL_RE = Pattern.compile("^(\\s*)(\\d+)([.)])\\s+(.*)$");
    private static final Pattern BQ_RE = Pattern.compile("^>\\s?(.*)$");
    private static final Pattern FENCE_RE = Pattern.compile("^(`{3,})\\s*(\\S*)\\s*$");
    private static final Pattern ROW_RE = Pattern.compile("^\\|(.+)\\|\\s*$");
    private static final Pattern DELIM_RE = Pattern.compile("^\\|\\s*:?-+:?\\s*(\\|\\s*:?-+:?\\s*)*\\|\\s*$");
    private static final Pattern HEADING_RE = Pattern.compile("^(#{1,6})\\s+(.*?)(?:\\s+#+)?\\s*$");
    private static final Pattern HEADING_PREFIX_RE = Pattern.compile("^#{1,6}\\s+");

    /** Scans {@code source} into an immutable-content list of {@link Block} nodes. */
    public static List<Block> scan(String source) {
        List<Block> blocks = new ArrayList<>();
        List<RawLine> lines = splitLines(source);

        int i = 0;
        while (i < lines.size()) {
            RawLine line = lines.get(i);
            if (line.text().trim().isEmpty()) {
                i++;
                continue;
            }

            FenceResult fence = tryFence(source, lines, i);
            if (fence != null) {
                blocks.add(fence.fence());
                i = fence.next();
                continue;
            }
            TableResult tableResult = tryTable(source, lines, i);
            if (tableResult != null) {
                blocks.add(tableResult.table());
                i = tableResult.next();
                continue;
            }
            Block thematic = tryThematicBreak(source, line);
            if (thematic != null) {
                blocks.add(thematic);
                i++;
                continue;
            }
            BlockquoteResult bqResult = tryBlockquote(source, lines, i);
            if (bqResult != null) {
                blocks.add(bqResult.quote());
                i = bqResult.next();
                continue;
            }
            Block heading = tryHeading(source, line);
            if (heading != null) {
                blocks.add(heading);
                i++;
                continue;
            }
            Block listItem = tryListItem(source, line);
            if (listItem != null) {
                blocks.add(listItem);
                i++;
                continue;
            }
            ParagraphResult paragraphResult = consumeParagraph(source, lines, i);
            blocks.add(paragraphResult.paragraph());
            i = paragraphResult.next();
        }
        return List.copyOf(blocks);
    }

    private static List<RawLine> splitLines(String source) {
        List<RawLine> out = new ArrayList<>();
        int start = 0;
        int n = source.length();
        for (int i = 0; i < n; i++) {
            if (source.charAt(i) == '\n') {
                out.add(new RawLine(source.substring(start, i), start, i));
                start = i + 1;
            }
        }
        if (start <= n) {
            out.add(new RawLine(source.substring(start), start, n));
        }
        return out;
    }

    private static Block tryThematicBreak(String source, RawLine line) {
        if (!THEMATIC_RE.matcher(line.text()).find()) return null;
        return new Ast.ThematicBreak(Span.spanFromOffsets(source, line.startOffset(), line.endOffset()));
    }

    private static Block tryHeading(String source, RawLine line) {
        Matcher m = HEADING_RE.matcher(line.text());
        if (!m.find()) return null;
        String hashes = m.group(1);
        String text = m.group(2).trim();
        int level = hashes.length();
        Span span = Span.spanFromOffsets(source, line.startOffset(), line.endOffset());
        return new Ast.Heading(level, text, span);
    }

    private static Block tryListItem(String source, RawLine line) {
        Matcher ul = UL_RE.matcher(line.text());
        if (ul.find()) {
            String text = ul.group(3);
            int markerStart = line.startOffset() + ul.group(1).length();
            int markerEnd = markerStart + ul.group(2).length();
            int textStart = line.startOffset() + line.text().indexOf(text);
            return new Ast.ListItem(
                    text,
                    Span.spanFromOffsets(source, line.startOffset(), line.endOffset()),
                    List.of(new SegmentOffset(0, textStart)),
                    false,
                    Span.spanFromOffsets(source, markerStart, markerEnd));
        }
        Matcher ol = OL_RE.matcher(line.text());
        if (ol.find()) {
            String text = ol.group(4);
            int markerStart = line.startOffset() + ol.group(1).length();
            int markerEnd = markerStart + ol.group(2).length() + ol.group(3).length();
            int textStart = line.startOffset() + line.text().indexOf(text);
            return new Ast.ListItem(
                    text,
                    Span.spanFromOffsets(source, line.startOffset(), line.endOffset()),
                    List.of(new SegmentOffset(0, textStart)),
                    true,
                    Span.spanFromOffsets(source, markerStart, markerEnd));
        }
        return null;
    }

    private record BlockquoteResult(Ast.Blockquote quote, int next) {}

    private static BlockquoteResult tryBlockquote(String source, List<RawLine> lines, int startIdx) {
        RawLine first = lines.get(startIdx);
        Matcher m = BQ_RE.matcher(first.text());
        if (!m.find()) return null;

        // Each quoted line drops its `> ` prefix — block structure, not text — so
        // the joined text needs one segment entry per line to map back to source.
        String firstSegment = m.group(1);
        List<String> segments = new ArrayList<>();
        segments.add(firstSegment);
        List<SegmentOffset> segmentMap = new ArrayList<>();
        segmentMap.add(new SegmentOffset(0, first.startOffset() + first.text().indexOf(firstSegment)));
        int joinedTextOffset = firstSegment.length();

        int i = startIdx + 1;
        int endOffset = first.endOffset();
        while (i < lines.size()) {
            RawLine ln = lines.get(i);
            Matcher next = BQ_RE.matcher(ln.text());
            if (!next.find()) break;
            String segment = next.group(1);
            joinedTextOffset += 1; // newline separator
            segmentMap.add(new SegmentOffset(
                    joinedTextOffset, ln.startOffset() + ln.text().indexOf(segment)));
            segments.add(segment);
            joinedTextOffset += segment.length();
            endOffset = ln.endOffset();
            i++;
        }
        Ast.Blockquote quote = new Ast.Blockquote(
                String.join("\n", segments), Span.spanFromOffsets(source, first.startOffset(), endOffset), segmentMap);
        return new BlockquoteResult(quote, i);
    }

    private record ParagraphResult(Ast.Paragraph paragraph, int next) {}

    private static ParagraphResult consumeParagraph(String source, List<RawLine> lines, int startIdx) {
        RawLine first = lines.get(startIdx);
        int endIdx = startIdx;
        while (endIdx + 1 < lines.size()) {
            int candidateIdx = endIdx + 1;
            RawLine candidate = lines.get(candidateIdx);
            if (candidate.text().trim().isEmpty()) break;
            if (HEADING_PREFIX_RE.matcher(candidate.text()).find()) break;
            if (UL_RE.matcher(candidate.text()).find()) break;
            if (OL_RE.matcher(candidate.text()).find()) break;
            if (BQ_RE.matcher(candidate.text()).find()) break;
            if (FENCE_RE.matcher(candidate.text()).find()) break;
            if (ROW_RE.matcher(candidate.text()).find()) break;
            if (THEMATIC_RE.matcher(candidate.text()).find()) break;
            endIdx++;
        }
        RawLine last = lines.get(endIdx);
        int startOffset = first.startOffset();
        int endOffset = last.endOffset();
        Ast.Paragraph paragraph = new Ast.Paragraph(
                source.substring(startOffset, endOffset),
                Span.spanFromOffsets(source, startOffset, endOffset),
                List.of(new SegmentOffset(0, startOffset)));
        return new ParagraphResult(paragraph, endIdx + 1);
    }

    private record FenceResult(Ast.Fence fence, int next) {}

    private static FenceResult tryFence(String source, List<RawLine> lines, int startIdx) {
        RawLine start = lines.get(startIdx);
        Matcher open = FENCE_RE.matcher(start.text());
        if (!open.find()) return null;
        String fenceMarker = open.group(1);
        String info = open.group(2).trim();

        int i = startIdx + 1;
        Integer bodyStart = null;
        Integer bodyEnd = null;
        int endOffset = start.endOffset();
        while (i < lines.size()) {
            RawLine ln = lines.get(i);
            Matcher close = FENCE_RE.matcher(ln.text());
            if (close.find() && close.group(1).length() >= fenceMarker.length()) {
                endOffset = ln.endOffset();
                break;
            }
            if (bodyStart == null) bodyStart = ln.startOffset();
            bodyEnd = ln.endOffset() + 1; // include the newline that separates from the next line
            i++;
        }
        // A trailing +1 on the last line's endOffset can overshoot source.length() when an
        // unclosed fence runs to end-of-input with no final newline. JS's String.slice clamps an
        // out-of-range end argument internally, so clamping here reproduces the same body TEXT JS
        // would produce (the full remaining tail, not an empty string). It does NOT reproduce JS's
        // bodySpan.endOffset/endCol for this case: JS's spanFromOffsets does not clamp its
        // endOffset argument, so JS's real (unclamped) span endpoint is an out-of-range value —
        // itself just an artifact of charCodeAt returning NaN for out-of-range access (never
        // equal to '\n', so the line/col loop increments once more), not a documented contract.
        // We deliberately don't chase that OOB quirk here: no conformance bundle exercises an
        // unclosed fence with no trailing newline, so this divergence has no observable effect
        // today, and replicating it would mean engineering a fragile OOB-quirk mechanism for a
        // bug nobody observes. Clamping avoids a StringIndexOutOfBoundsException; nothing more.
        int clampedBodyEnd = bodyEnd == null ? 0 : Math.min(bodyEnd, source.length());
        String body = (bodyStart != null && bodyEnd != null) ? source.substring(bodyStart, clampedBodyEnd) : "";
        int fallbackOffset = start.endOffset();
        Span bodySpan = Span.spanFromOffsets(
                source,
                bodyStart != null ? bodyStart : fallbackOffset,
                bodyEnd != null ? clampedBodyEnd : fallbackOffset);
        Ast.Fence fenceBlock =
                new Ast.Fence(Span.spanFromOffsets(source, start.startOffset(), endOffset), info, body, bodySpan);
        return new FenceResult(fenceBlock, i + 1);
    }

    private record TableResult(Ast.Table table, int next) {}

    private static TableResult tryTable(String source, List<RawLine> lines, int startIdx) {
        if (startIdx + 1 >= lines.size()) return null;
        RawLine headerLine = lines.get(startIdx);
        RawLine delimLine = lines.get(startIdx + 1);
        if (!ROW_RE.matcher(headerLine.text()).find()) return null;
        if (!DELIM_RE.matcher(delimLine.text()).find()) return null;

        RowCells headerParsed = TableCells.parseRowCells(headerLine.text(), headerLine.startOffset(), source);
        Row header = new Row(
                headerParsed.cells(),
                headerParsed.cellSpans(),
                Span.spanFromOffsets(source, headerLine.startOffset(), headerLine.endOffset()));

        List<Row> rows = new ArrayList<>();
        int i = startIdx + 2;
        while (i < lines.size()) {
            RawLine ln = lines.get(i);
            if (!ROW_RE.matcher(ln.text()).find()) break;
            RowCells parsed = TableCells.parseRowCells(ln.text(), ln.startOffset(), source);
            rows.add(new Row(
                    parsed.cells(),
                    parsed.cellSpans(),
                    Span.spanFromOffsets(source, ln.startOffset(), ln.endOffset())));
            i++;
        }
        Row lastRow = rows.isEmpty() ? null : rows.get(rows.size() - 1);
        int endOffset = lastRow != null ? lastRow.span().endOffset() : delimLine.endOffset();
        Ast.Table table =
                new Ast.Table(Span.spanFromOffsets(source, headerLine.startOffset(), endOffset), header, rows);
        return new TableResult(table, i);
    }
}

File: typescript/packages/core/src/scanner.ts
import type { Block, SegmentOffset } from './ast.ts'
import { type Span, spanFromOffsets } from './span.ts'
import { parseRowCells } from './table-cells.ts'

// A scanned source line with the offsets needed to materialize spans on the
// blocks built from it.
type RawLine = {
  readonly text: string
  readonly startOffset: number
  readonly endOffset: number
}

export function scan(source: string): ReadonlyArray<Block> {
  const blocks: Block[] = []
  const lines = splitLines(source)

  let i = 0
  while (i < lines.length) {
    const line = lines[i]
    if (!line) {
      i++
      continue
    }
    if (line.text.trim().length === 0) {
      i++
      continue
    }
    const fence = tryFence(source, lines, i)
    if (fence) {
      blocks.push(fence.fence)
      i = fence.next
      continue
    }
    const tableResult = tryTable(source, lines, i)
    if (tableResult) {
      blocks.push(tableResult.table)
      i = tableResult.next
      continue
    }
    const thematic = tryThematic(source, line)
    if (thematic) {
      blocks.push(thematic)
      i++
      continue
    }
    const bqResult = tryBlockquote(source, lines, i)
    if (bqResult) {
      blocks.push(bqResult.quote)
      i = bqResult.next
      continue
    }
    const heading = tryHeading(source, line)
    if (heading) {
      blocks.push(heading)
      i++
      continue
    }
    const listItem = tryListItem(source, line)
    if (listItem) {
      blocks.push(listItem)
      i++
      continue
    }
    const { paragraph, next } = consumeParagraph(source, lines, i)
    blocks.push(paragraph)
    i = next
  }
  return blocks
}

function splitLines(source: string): ReadonlyArray<RawLine> {
  const out: RawLine[] = []
  let start = 0
  for (let i = 0; i < source.length; i++) {
    if (source.charCodeAt(i) === 0x0a) {
      out.push({ text: source.slice(start, i), startOffset: start, endOffset: i })
      start = i + 1
    }
  }
  if (start <= source.length) {
    out.push({ text: source.slice(start), startOffset: start, endOffset: source.length })
  }
  return out
}

const THEMATIC_RE = /^\s*([-*_])(\s*\1){2,}\s*$/
const UL_RE = /^(\s*)([-*+])\s+(.*)$/
const OL_RE = /^(\s*)(\d+)([.)])\s+(.*)$/
const BQ_RE = /^>\s?(.*)$/

function tryThematic(source: string, line: RawLine): Block | undefined {
  if (!THEMATIC_RE.test(line.text)) return undefined
  return {
    kind: 'thematic_break',
    span: spanFromOffsets(source, line.startOffset, line.endOffset),
  }
}

function tryHeading(source: string, line: RawLine): Block | undefined {
  const m = /^(#{1,6})\s+(.*?)(?:\s+#+)?\s*$/.exec(line.text)
  if (!m) return undefined
  const hashes = m[1] ?? ''
  const text = (m[2] ?? '').trim()
  const level = hashes.length as 1 | 2 | 3 | 4 | 5 | 6
  const span = spanFromOffsets(source, line.startOffset, line.endOffset)
  return { kind: 'heading', level, text, span }
}

function tryListItem(source: string, line: RawLine): Block | undefined {
  const ul = UL_RE.exec(line.text)
  if (ul) {
    const text = ul[3] ?? ''
    const markerStart = line.startOffset + (ul[1] ?? '').length
    const markerEnd = markerStart + (ul[2] ?? '').length
    const textStart = line.startOffset + line.text.indexOf(text)
    return {
      kind: 'list_item',
      ordered: false,
      text,
      span: spanFromOffsets(source, line.startOffset, line.endOffset),
      segmentMap: [{ textOffset: 0, sourceOffset: textStart }],
      markerSpan: spanFromOffsets(source, markerStart, markerEnd),
    }
  }
  const ol = OL_RE.exec(line.text)
  if (ol) {
    const text = ol[4] ?? ''
    const markerStart = line.startOffset + (ol[1] ?? '').length
    const markerEnd = markerStart + (ol[2] ?? '').length + (ol[3] ?? '').length
    const textStart = line.startOffset + line.text.indexOf(text)
    return {
      kind: 'list_item',
      ordered: true,
      text,
      span: spanFromOffsets(source, line.startOffset, line.endOffset),
      segmentMap: [{ textOffset: 0, sourceOffset: textStart }],
      markerSpan: spanFromOffsets(source, markerStart, markerEnd),
    }
  }
  return undefined
}

function tryBlockquote(
  source: string,
  lines: ReadonlyArray<RawLine>,
  startIdx: number,
): { quote: Block; next: number } | undefined {
  const first = lines[startIdx]
  if (!first) return undefined
  const m = BQ_RE.exec(first.text)
  if (!m) return undefined

  // Each quoted line drops its `> ` prefix — block structure, not text — so
  // the joined text needs one segment entry per line to map back to source.
  const firstSegment = m[1] ?? ''
  const segments: string[] = [firstSegment]
  const segmentMap: SegmentOffset[] = [
    { textOffset: 0, sourceOffset: first.startOffset + first.text.indexOf(firstSegment) },
  ]
  let joinedTextOffset = firstSegment.length

  let i = startIdx + 1
  let endOffset = first.endOffset
  while (i < lines.length) {
    const ln = lines[i]
    if (!ln) break
    const next = BQ_RE.exec(ln.text)
    if (!next) break
    const segment = next[1] ?? ''
    joinedTextOffset += 1 // newline separator
    segmentMap.push({
      textOffset: joinedTextOffset,
      sourceOffset: ln.startOffset + ln.text.indexOf(segment),
    })
    segments.push(segment)
    joinedTextOffset += segment.length
    endOffset = ln.endOffset
    i++
  }
  return {
    quote: {
      kind: 'blockquote',
      text: segments.join('\n'),
      span: spanFromOffsets(source, first.startOffset, endOffset),
      segmentMap,
    },
    next: i,
  }
}

function consumeParagraph(
  source: string,
  lines: ReadonlyArray<RawLine>,
  startIdx: number,
): { paragraph: Block; next: number } {
  const first = lines[startIdx]
  if (!first) throw new Error('invariant: startIdx out of range')
  let endIdx = startIdx
  while (endIdx + 1 < lines.length) {
    const candidateIdx = endIdx + 1
    const candidate = lines[candidateIdx]
    if (!candidate) break
    if (candidate.text.trim().length === 0) break
    if (/^#{1,6}\s+/.test(candidate.text)) break
    if (UL_RE.test(candidate.text)) break
    if (OL_RE.test(candidate.text)) break
    if (BQ_RE.test(candidate.text)) break
    if (FENCE_RE.test(candidate.text)) break
    if (ROW_RE.test(candidate.text)) break
    if (THEMATIC_RE.test(candidate.text)) break
    endIdx++
  }
  const last = lines[endIdx]
  if (!last) throw new Error('invariant: endIdx out of range')
  const startOffset = first.startOffset
  const endOffset = last.endOffset
  return {
    paragraph: {
      kind: 'paragraph',
      text: source.slice(startOffset, endOffset),
      span: spanFromOffsets(source, startOffset, endOffset),
      segmentMap: [{ textOffset: 0, sourceOffset: startOffset }],
    },
    next: endIdx + 1,
  }
}

const FENCE_RE = /^(`{3,})\s*(\S*)\s*$/

function tryFence(
  source: string,
  lines: ReadonlyArray<RawLine>,
  startIdx: number,
): { fence: Block; next: number } | undefined {
  const start = lines[startIdx]
  if (!start) return undefined
  const open = FENCE_RE.exec(start.text)
  if (!open) return undefined
  const fenceMarker = open[1] ?? ''
  const info = (open[2] ?? '').trim()
  let i = startIdx + 1
  let bodyStart: number | undefined
  let bodyEnd: number | undefined
  let endOffset = start.endOffset
  while (i < lines.length) {
    const ln = lines[i]
    if (!ln) {
      i++
      continue
    }
    const close = FENCE_RE.exec(ln.text)
    if (close && (close[1] ?? '').length >= fenceMarker.length) {
      endOffset = ln.endOffset
      break
    }
    if (bodyStart === undefined) bodyStart = ln.startOffset
    bodyEnd = ln.endOffset + 1 /* include the newline that separates from next line */
    i++
  }
  const body =
    bodyStart !== undefined && bodyEnd !== undefined ? source.slice(bodyStart, bodyEnd) : ''
  const bodySpan = spanFromOffsets(source, bodyStart ?? start.endOffset, bodyEnd ?? start.endOffset)
  return {
    fence: {
      kind: 'fence',
      info,
      body,
      bodySpan,
      span: spanFromOffsets(source, start.startOffset, endOffset),
    },
    next: i + 1,
  }
}

const ROW_RE = /^\|(.+)\|\s*$/
const DELIM_RE = /^\|\s*:?-+:?\s*(\|\s*:?-+:?\s*)*\|\s*$/

function tryTable(
  source: string,
  lines: ReadonlyArray<RawLine>,
  startIdx: number,
): { table: Block; next: number } | undefined {
  const headerLine = lines[startIdx]
  const delimLine = lines[startIdx + 1]
  if (!headerLine || !delimLine) return undefined
  if (!ROW_RE.test(headerLine.text)) return undefined
  if (!DELIM_RE.test(delimLine.text)) return undefined
  const headerParsed = parseRowCells(headerLine.text, headerLine.startOffset, source)
  const header = {
    cells: headerParsed.cells,
    cellSpans: headerParsed.cellSpans,
    span: spanFromOffsets(source, headerLine.startOffset, headerLine.endOffset),
  }
  const rows: {
    cells: ReadonlyArray<string>
    cellSpans: ReadonlyArray<Span>
    span: Span
  }[] = []
  let i = startIdx + 2
  while (i < lines.length) {
    const ln = lines[i]
    if (!ln) break
    if (!ROW_RE.test(ln.text)) break
    const parsed = parseRowCells(ln.text, ln.startOffset, source)
    rows.push({
      cells: parsed.cells,
      cellSpans: parsed.cellSpans,
      span: spanFromOffsets(source, ln.startOffset, ln.endOffset),
    })
    i++
  }
  const lastRow = rows[rows.length - 1]
  const endOffset = lastRow ? lastRow.span.endOffset : delimLine.endOffset
  return {
    table: {
      kind: 'table',
      span: spanFromOffsets(source, headerLine.startOffset, endOffset),
      header,
      rows,
    },
    next: i,
  }
}
